gapir/cc/interpreter.cpp

```diff
@@ -236,27 +337,82 @@ Interpreter::Result Interpreter::pushI(uint32_t opcode) {
     GAPID_WARNING("Error: pushI basic type invalid %d", (int)type);
     return ERROR;
   }
-  Stack::BaseValue data = extract20bitData(opcode);
+
+  uint64_t data = extract20bitData(opcode);
+
   switch (type) {
-    // Sign extension for signed types
+    case BaseType::Bool:
+      mStack.push(data != 0);
+      break;
+
+    case BaseType::Int8:
+      mStack.push<int8_t>(data);
+      break;
+
+    case BaseType::Int16:
+      mStack.push<int16_t>(data);
+      break;
+
     case BaseType::Int32:
+      data |= (data & 0x80000) != 0
+                  ? 0xfff00000ULL
+                  : 0x0;  // Sign extend the 20 bit immediate value
+      mStack.push(*((int32_t*)&data));
```

Comment:
This breaks strict aliasing rules.

Reply:
Do we not break aliasing rules all over GAPID with all the void* trickery we're performing? The whole conceptual operation of this function implies type aliasing.

Reply:
We do it in a slightly different way for the old system, but I guess we run into similar problems with the union.
The spec-compliant way to solve this is with a memcpy
```
int32_t v;
memcpy(&v, &data, sizeof(v));
mStack.push(v);
```

A memcpy is probably also the right thing to do in the set functions. As long as the compile optimizes memcpy (which many do) then it should remain valid.